subprojects/platform-native/src/main/java/org/gradle/nativeplatform/internal/services/NativeBinaryServices.java

```diff
@@ -64,6 +73,26 @@ public void registerProjectServices(ServiceRegistration registration) {
         WindowsSdkLocator createWindowsSdkLocator(OperatingSystem os, WindowsRegistry windowsRegistry) {
             return new DefaultWindowsSdkLocator(os, windowsRegistry);
         }
+
+        VisualCppMetadataProvider createVisualCppMetadataProvider(WindowsRegistry windowsRegistry) {
+            return new DefaultVisualCppMetadataProvider(windowsRegistry);
+        }
+
+        WindowsRegistryVersionLocator createWindowsRegistryVersionLocator(WindowsRegistry windowsRegistry) {
+            return new WindowsRegistryVersionLocator(windowsRegistry);
+        }
+
+        CommandLineVersionLocator createCommandLineVersionLocator(ExecActionFactory execActionFactory, WindowsRegistry windowsRegistry, OperatingSystem os, VisualCppMetadataProvider visualCppMetadataProvider) {
+            return new CommandLineVersionLocator(execActionFactory, windowsRegistry, os, visualCppMetadataProvider);
+        }
+
```

Comment:
Would it make sense to make the locators build session/build tree scoped so we only need to do the "expensive" look-ups once per session/composite?

Reply:
Probably.  I've moved them up.